### DISCUSSION TITLE
Add Laravel Octane Support

### DISCUSSION THREAD

**Discussion by jaydrogers** [serversideup/docker-php#72] [Ideas & Feature Requests]
# Background
* Swoole is a newish technology in the PHP world
* Laravel just came out with "Octane" for high performance support: https://laravel.com/docs/9.x/octane

# ✅  UDPATE: Laravel Octane now available!
Laravel Octane support is now natively available with [v4.0.0](https://github.com/serversideup/docker-php/releases/tag/v4.0.0) 🎉

[Learn more about Laravel Octane →](https://serversideup.net/open-source/docker-php/docs/framework-guides/laravel/octane)

**Comment by jaydrogers** [comment]
Just adding a note that @fideloper added a PR for one of his projects using Octane. This might be a good thing to reference and bring back into the project natively: https://github.com/superfly/flyctl/pull/1143

**Comment by iksaku** [comment]
I'm running Octane in production. At the moment I'm using RoadRunner, but Swoole is also good!

Due to personal preference, I ended up [building my own images in Alpine](https://github.com/iksaku/laravel-alpine-images), but it may be useful for your research.

Let me know if there's anything I could help with and I will try to answer to the best of my abilities 🙌

**Comment by wilsenhc** [comment]
Having support for Octane with Swoole would be awesome.

Thanks for the good work @jaydrogers 🙌🏼

**Comment by jihadismail8** [comment]
that would be nice to do , i  would love to have alpine version of laravel octane on swoole .

**Comment by anotherglitchinthematrix** [comment]
We're waiting for this :^), already using the docker-php@2.x and very satified

**Comment by ChiefC0de** [comment]
Up! Hope you will add such feature.. We really need it!

**Comment by Sammyjo20** [comment]
Just putting another post in here to say that this would be an amazing addition!

**Reply by ChiefC0de** [reply]
@Sammyjo20 seems they dont really care about it. Use official frankenphp docker image itself, thats all I can say. It covers all things, you dont even need nginx.

**Reply by ChiefC0de** [reply]
Just check their official docs

**Reply by ChiefC0de** [reply]
Or you can check exaco/laravel-octane-dockerfile . At least it always gets updated

**Reply by Sammyjo20** [reply]
I don't think it's a case that they don't care about implementing it - they have put a lot of effort into serversideup/spin which is probably their preference, plus it's not really meant to be Laravel specific which is why it's probably not out yet. I'm going to see if I can get it running with the Frankenphp docker image or maybe even running Octane as a task with serversideup/php like the scheduler/horizon is configured.

**Reply by jaydrogers** [reply]
> @Sammyjo20 seems they dont really care about it.

Yeah... came here to address this comment.

We care about it, but we've just had 575+ hours of open source work to get to before we've had a chance to get to this (while still taking on client work to pay the bills).

<img width="982" alt="image" src="https://github.com/serversideup/docker-php/assets/3174134/c3dbc8be-d993-43cd-b4f1-093ab1b4e29f">

@ChiefC0de: Moving forward, please treat others with respect before posting assumptions like this. Other open source maintainers may not have an understanding that you didn't intend to insult others in the community.

**Reply by ChiefC0de** [reply]
I don't disrespect your work. I understand how hard it is to put your hours for open source while still trying to pay the bills. But this whole problem with octane wasn't solved yet and I had a lot of problems while using your images and trying to make it work and spent a lot of time looking for your repo and still you cant implement this - just facts, I dont try to offend you. Sorry if I did. Hope you will get it done one day and I will come back. But for now - it's not ready for the real production. Still wish you all the best and hope to see it done. Thanks for understanding. Good luck.

**Reply by jaydrogers** [reply]
@ChiefC0de: We're happy and willing to accept any PRs that help bring the community this feature 👍

If no PRs are submitted, we'll just have to wait until the maintainers have time to get to it. It's not a feature that we've forgotten about.

**Comment by spaceworkplatform** [comment]
any news/plans about this ? we really want to use spin pro but we relay on Laravel Octane.

**Comment by balthild** [comment]
FYI, this is my Dockerfile that runs octane in `*-fpm-nginx` images with minimal modification to the container.

```dockerfile
FROM serversideup/php:8.4-fpm-nginx

USER root
RUN install-php-extensions openswoole
RUN sed -i 's|try_files .*;|try_files $uri $uri/ @octane;|g' /etc/nginx/site-opts.d/*.conf.template
RUN sed -i 's|fastcgi_pass .*;|try_files /.well-known/dummy @octane;|g' /etc/nginx/site-opts.d/*.conf.template
RUN printf SIGINT > /etc/s6-overlay/s6-rc.d/php-fpm/down-signal
USER www-data

COPY --chmod=0755 --chown=root:root <<'EOF' /etc/s6-overlay/s6-rc.d/php-fpm/run
#!/command/execlineb -P
with-contenv
s6-notifyoncheck -d
/usr/local/bin/php /var/www/html/artisan octane:start --port=9000
EOF

COPY --chmod=0755 --chown=root:root <<'EOF' /usr/local/bin/php-fpm-healthcheck
#!/bin/sh
/usr/local/bin/php /var/www/html/artisan octane:status
EOF

COPY --chmod=0755 --chown=root:root <<'EOF' /usr/local/bin/nginx-healthcheck
#!/bin/sh
curl --insecure --silent --location --show-error --fail http://localhost:8080$HEALTHCHECK_PATH
EOF

COPY --chown=www-data:www-data <<'EOF' /etc/nginx/conf.d/connection_upgrade.conf
map $http_upgrade $connection_upgrade {
    default upgrade;
    ''      close;
}
EOF

COPY --chown=www-data:www-data <<'EOF' /etc/nginx/server-opts.d/octane.conf
location @octane {
    set $suffix "";

    if ($uri = /index.php) {
        set $suffix ?$query_string;
    }

    proxy_http_version 1.1;
    proxy_set_header Host $http_host;
    proxy_set_header Scheme $scheme;
    proxy_set_header SERVER_PORT $server_port;
    proxy_set_header REMOTE_ADDR $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $connection_upgrade;

    proxy_pass http://127.0.0.1:9000$suffix;
}
EOF

COPY --chown=www-data:www-data <<'EOF' /etc/nginx/server-opts.d/healthcheck-octane.conf
location /healthcheck-octane {
    access_log off;
    return 200 "OK";
}
EOF

ENV HEALTHCHECK_PATH="/healthcheck-octane"
HEALTHCHECK --interval=5s --timeout=3s --retries=3 \
    CMD [ "sh", "-c", "nginx-healthcheck && php-fpm-healthcheck || exit 1" ]
```

Looks like it would be much easier with `*-unit` variations, but we're not currently using nginx unit now.

**Reply by jaydrogers** [reply]
This is awesome! Thanks for sharing.

I appreciate sharing what you have. I will work to take these concepts and make it easier to enable without needing configuration modifications 👍

**Comment by jaydrogers** [comment]
# 🚀 Huge update
Native Octane support is now in beta: https://github.com/serversideup/docker-php/releases/tag/v4.0.0-beta3

**Comment by Tamas-hi** [comment]
Wow! Just found about this project today. The whole 4.0 release looks so professional, the docs explains better what docker/frankenphp/octane is than their official docs and I am just stunned. Definitely the way forward in our ecosystem!

**Reply by jaydrogers** [reply]
Thank you! I greatly appreciate your kind words 😊

**Comment by jaydrogers** [comment]
# ✅ Laravel Octane now available!
Laravel Octane is now natively supported with [v4.0.0](https://github.com/serversideup/docker-php/releases/tag/v4.0.0) 🎉

[Learn more about Laravel Octane →](https://serversideup.net/open-source/docker-php/docs/framework-guides/laravel/octane)

**Comment by ronald2wing** [comment]
Can we make the octane setup cleaner by using ENV and embed the exec php artisan octane:start --server=frankenphp --port=8080
in AUTORUN scripts?
